File: B3WM/B3WM.Client/Model/VolumeLevelStorageItem.cs
using B3WM.Shared.Entity;
using Magic.IndexedDb.SchemaAnnotations;
using Nextended.Core.Extensions;
using System.Collections.Concurrent;
using System.Text.Json.Serialization;

namespace B3WM.Client.Model
{
    /// <summary>
    /// VolumeLevel para IndexedDB com data da barra. KeyPath "id" (case-sensitive).
    /// BarDate permite filtrar volume por intervalo de datas ao remontar.
    /// </summary>
    public class VolumeLevelStorageItem
    {
        public int Id { get; set; }

        public DateTime Date { get; set; }

        public string Symbol { get; set; }

        //InMinutes
        public int TimeFrame { get; set; }

        public List<VolumeLevel> Volumes { get; set; } = new();

        public static List<VolumeLevel> FromTo(List<BarStorageItem> data, DateTime from, DateTime to)
        {
            List<VolumeLevel> result = new List<VolumeLevel>();

            Console.WriteLine($"From: {from}, To: {to}");

            var barStorageTo = data.FirstOrDefault(q => q.Date == to);
            var barStorageFrom = data.FirstOrDefault(q => q.Date == from);

            if (barStorageTo == null || barStorageFrom == null) return new();

            result = barStorageTo.VolumeLevel ?? new();

            var grouping = data.GroupBy(q => q.Date.Date);

            //sum last register of volumes
            foreach(var item in grouping.Where(q=>q.Key >= from.Date && q.Key < to.Date).OrderByDescending(q=>q.Key))
            {
                //get last voluems
                var volumeLevel = item.OrderByDescending(q=>q.Date).First().VolumeLevel;

                result = Operation(result, volumeLevel ?? new(), "Sum");
            }

            result = Operation(result, barStorageFrom.VolumeLevel ?? new(), "Diff");


            return result.OrderBy(q=>q.Price).ToList();
        }

        public static List<VolumeLevel> Operation(List<VolumeLevel> vol1, List<VolumeLevel> vol2, string Operation)
        {
            ConcurrentDictionary<double, VolumeLevel> dictionary = new ConcurrentDictionary<double, VolumeLevel>();

            foreach(var item in vol1)
            {
                var itemVol2 = vol2.FirstOrDefault(q => q.Price == item.Price);

                dictionary.TryAdd(item.Price, new VolumeLevel()
                {
                    Total = item.Total - (itemVol2?.Total ?? 0),
                    BuyVolume = item.BuyVolume - (itemVol2?.BuyVolume ?? 0),
                    SellVolume = item.SellVolume - (itemVol2?.SellVolume ?? 0),
                    Price = item.Price
                });
            }

            return dictionary.Select(e=>e.Value).ToList();
        }
    }
}

File: B3WM.Flutter/lib/models/volume_level_storage_item.dart
import 'volume_level.dart';
import 'bar_storage_item.dart';

class VolumeLevelStorageItem {
  final int id;
  final DateTime date;
  final String symbol;
  final int timeFrame;
  final List<VolumeLevel> volumes;

  VolumeLevelStorageItem({
    required this.id,
    required this.date,
    required this.symbol,
    required this.timeFrame,
    required this.volumes,
  });

  factory VolumeLevelStorageItem.fromJson(Map<String, dynamic> json) {
    return VolumeLevelStorageItem(
      id: json['id'] as int? ?? 0,
      date: DateTime.parse(json['date'] as String),
      symbol: json['symbol'] as String? ?? '',
      timeFrame: json['timeFrame'] as int? ?? 0,
      volumes: (json['volumes'] as List?)
              ?.map((e) => VolumeLevel.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'date': date.toIso8601String(),
        'symbol': symbol,
        'timeFrame': timeFrame,
        'volumes': volumes.map((e) => e.toJson()).toList(),
      };

  static List<VolumeLevel> fromTo(
      List<BarStorageItem> data, DateTime from, DateTime to) {
    final barTo = data.where((b) => b.date == to).firstOrNull;
    final barFrom = data.where((b) => b.date == from).firstOrNull;
    if (barTo == null || barFrom == null) return [];

    DateTime dateOnly(DateTime dt) => DateTime(dt.year, dt.month, dt.day);
    var result = barTo.volumeLevel ?? <VolumeLevel>[];
    final fd = dateOnly(from);
    final td = dateOnly(to);
    final grouped = data
        .where((b) {
          final bd = dateOnly(b.date);
          return !bd.isBefore(fd) && bd.isBefore(td);
        })
        .groupBy((b) => dateOnly(b.date));

    for (final entry in grouped.entries) {
      final lastEntry = entry.value
          .reduce((a, b) => a.date.compareTo(b.date) > 0 ? a : b);
      if (lastEntry.volumeLevel != null) {
        result = operation(result, lastEntry.volumeLevel!, 'Sum');
      }
    }

    result = operation(result, barFrom.volumeLevel ?? [], 'Diff');
    result.sort((a, b) => a.price.compareTo(b.price));
    return result;
  }

  static List<VolumeLevel> operation(
      List<VolumeLevel> vol1, List<VolumeLevel> vol2, String op) {
    final map = <double, VolumeLevel>{};
    for (final item in vol1) {
      final match = vol2.where((v) => v.price == item.price).firstOrNull;
      if (op == 'Sum') {
        map[item.price] = VolumeLevel(
          total: item.total + (match?.total ?? 0),
          buyVolume: item.buyVolume + (match?.buyVolume ?? 0),
          sellVolume: item.sellVolume + (match?.sellVolume ?? 0),
          price: item.price,
        );
      } else {
        map[item.price] = VolumeLevel(
          total: item.total - (match?.total ?? 0),
          buyVolume: item.buyVolume - (match?.buyVolume ?? 0),
          sellVolume: item.sellVolume - (match?.sellVolume ?? 0),
          price: item.price,
        );
      }
    }
    return map.values.toList();
  }
}

extension _ListGroupBy<K, V> on Iterable<V> {
  Map<K, List<V>> groupBy(K Function(V) keyFn) {
    final map = <K, List<V>>{};
    for (final item in this) {
      final key = keyFn(item);
      map.putIfAbsent(key, () => []).add(item);
    }
    return map;
  }
}
